### PR TITLE
Upgrade recalbox.conf : 2nd attempt

### PR DESCRIPTION
**Pull request by substring**
Please make sure your PR is ready to be merged !
- [ ] You added the changes in CHANGELOG.md
- [x] You choose the right repository branch to make the PR
- [x] You described the PR as below

Fixes #XXXX

Changes :
This version is now a standalone function in a new .sh
Now it can :
- copy over protected values (like in the wifi password)
- append user defined properties that don't exist by default
- logs in recalbox.log
- property values can have a = inside (like for gamecon), still works
  Only drawback : takes about 7s on my pi3. Well, that's once every upgrade, people on the unstable branch will have to cope with it

Please comment and review

### CODE REVIEW

File: board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+# Upgrade the recalbox.conf if necessary
+function doRbxConfUpgrade {
+  # Update recalbox.conf
+  cfgIn=$_SHAREINIT/system/recalbox.conf
+  cfgOut=$_SHARE/system/recalbox.conf
+  rbxVersion=$_RBX/recalbox.version
+  curVersion=$_SHARE/system/logs/lastrecalbox.conf.update
+  updVersion=$_RBX/recalbox.updateversion
+  forced=(controllers.ps3.driver)
+  savefile=${cfgOut}-pre-$(cat $updVersion)
```
Comment (digitalLumberjack) [inline board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh]: savefile=${cfgOut}-pre-$(cat $rbxVersion)

File: board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+# Upgrade the recalbox.conf if necessary
+function doRbxConfUpgrade {
+  # Update recalbox.conf
+  cfgIn=$_SHAREINIT/system/recalbox.conf
+  cfgOut=$_SHARE/system/recalbox.conf
+  rbxVersion=$_RBX/recalbox.version
+  curVersion=$_SHARE/system/logs/lastrecalbox.conf.update
+  updVersion=$_RBX/recalbox.updateversion
```
Comment (digitalLumberjack) [inline board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh]: This file does not exist.
We already are on the new version as the update has already be done.
We can remove this file.

**Review by digitalLumberjack — CHANGES_REQUESTED** [review]
Nearly perfect :)

File: board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh (hunk 1 above)
Comment (substring) [reply board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh]: would need some string mnipulation as it has afull datetime in it ( some / and : )
In that case, the curVersion var is not necessary either

Comment (digitalLumberjack) [reply board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh]: right

**Comment by substring** [comment]
@digitalLumberjack Changes made as requested. I preferred keeping the `~/logs/logs/lastrecalbox.conf.update` file and the `$curVersion` if the users decided to remove recalbox.conf backups. At least, that way, we manage in a different place the recalbox.conf version

File: board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  # Add each line of the share_init recalbox.conf, and decide what happens if a property is forced, custom or 
+  while read -r property ; do
+    name=$(echo $property | cut -d '=' -f 1)
+    realname=$(echo $name | sed 's/^;//g')
+
+    # Echo empty lines for \n
+    if [[ $property == "" ]]; then
+      echo "" >> $cfgOut
+    # Should we force the value ?
+    elif containsElement "$name" $forced; then
```
Comment (digitalLumberjack) [inline board/recalbox/fsoverlay/recalbox/scripts/recalbox-includes.sh]: elif containsElement "$name" "${forced[@]}"; then
Last effort :)

**Review by digitalLumberjack — CHANGES_REQUESTED** [review]
Just a little change

**Comment by substring** [comment]
@digitalLumberjack Requested changes were made + slight little improvements. See the commit for details

**Review by digitalLumberjack — APPROVED** [review]
LGTM

**Comment by substring** [comment]
@nadenislamarre your review would be appreciated pal :)

**Comment by nadenislamarre** [comment]
sorry, i'm not fan of your shell style as already told.
i've written quickly a similar one in shell, so you can inspire it to update yours, as you prefer.
you can run my example with : ./updateRecalboxConf.sh recalbox.conf.old recalbox.conf

[updaterecalbox.tar.gz](https://github.com/recalbox/recalbox-buildroot/files/509340/updaterecalbox.tar.gz)

**Comment by substring** [comment]
I like :
- the easier steps (copy the shareinit file to a temp file, look for keys from the share and rewrite them in the temp file)
- much faster
- error handling
- less arrays (which is a consequence of the 1st point)

I don't like :
- you expect some indentation with `[ ]*` as well as some spaces around the `=`.
- you're not using an array for forced values, it would make things easier to maintain and parse
- I don't see the aim of the LAST_KEY. If it's to make a difference in versions, there are much better options
- doesn't replace inline properties that were uncommented, it just writes them at the end of the file (ex : set some wifi SSID and PASS, they are appended instead of replaced)
- missing some logging (ok, that's a quick and dirty version)
- paranoid mode : exits on any error
- it doesn't even make coffe

@digitalLumberjack shall I try to make the best of the 2 scripts ?

**Comment by nadenislamarre** [comment]
> > you expect some indentation with [ ]\* as well as some spaces around the =.
> > i've not understood. Feel free to put the regular expression that works.
> > 
> > you're not using an array for forced values, it would make things easier to maintain and parse
> > keep it simple.
> > why make something complicate for something that will probably never evolve ?
> > more over, in case it evolves, we can change it. Note that if it evolves in a sense that regular expression make more sens, like ps3.*, it can be easier to maintain.
> > I prefer to make things more complex when they really happend.
> > 
> >    I don't see the aim of the LAST_KEY. If it's to make a difference in versions, there are much better options
> > you didn't want my revision property ;-( so it's just a quick way to check that the file is up to date. When you add values into recalbox.conf, you've to update this value : it's the drawback compared to the revision property.
> > 
> >    doesn't replace inline properties that were uncommented, it just writes them at the end of the file (ex : set some wifi SSID and PASS, they are appended instead of replaced)
> > in fact, i've just tested the file i put in example. So, possibly i don't handle some cases.
> > 
> >    paranoid mode : exits on any error
> > if you check what can cause the error, it's only disk errors or / in values that makes sed fails (because i've not escape them).
> > I don't find it paranoid, just safe.
> > 
> >    it doesn't even make coffe
> > yes, this was the main target of this script : show what a non java style shell script looks like ;-)

**Pull request closed without merging**